### PR TITLE
Allow nested conditionals to be related via constraints

### LINKED ISSUE #30706: NonNullable<ConditionalType<T>> can't be assigned to ConditionalType<T> with --strictNullChecks
**TypeScript Version:**  typescript@3.4.0-dev.20190330

**Search Terms: Conditional NonNullable Type**

**Code**

```ts
type ConditionalType<T> = T extends string ? string : number ;

function ConditionalOrUndefined<T>() : ConditionalType<T> | undefined {
    return 0 as any;
}

function JustConditional<T>() : ConditionalType<T> {
    return ConditionalOrUndefined<T>()!; // Type 'NonNullable<ConditionalType<T>>' is not assignable to type 'ConditionalType<T>'
}


// For comparison...
function genericOrUndefined<T>() : T | undefined {
    return 0 as any;
}

function JustGeneric<T>(): T {
    return genericOrUndefined<T>()!; // no error
}
```

**Expected behavior:**
JustConditional<T> compiles without errors, like JustGeneric<T>.

**Actual behavior:**
JustConditional produces error ``` Type 'NonNullable<ConditionalType<T>>' is not assignable to type 'ConditionalType<T>'```

**Playground Link:** 
https://typescript-play.js.org/?target=2#code/C4TwDgpgBAwg9gOwCYEtgsQQwDYBVwQA8uAfFALxS5QQAewEyAzlE8AE4oIDmUA-Kw5deALigIArgFsARhHZQA3AChlAMwkIAxukSxEqXQhwB5dgFVkENVwhJiJABQBKKGPjI0GY3gIOoAD5QmkjWtkhQAN7KULFQ7BDAEuwIUAAMUJgsmAggKgC+qhraRlAAUhJsHobeOA4ubvqeRjj4kP7RcfGJyanVXljYZpahNgh29c4AhIpQAPRzVARQAOQAcohrEtjYmDLYRP0tvu2kJCtQKCwIcMCZTEwo3Mb70MBwUKCQq0e1J0SkFbKQqqBZQABicAUWjgUjAmE4TEQADpUUVNDpvFBuIx5CgtMMrGMJqQGmJqEEQmFxhFOnEEkkUul7plcgV0SUsRU2ABxXGcLSTclRGL0npMnHjAWE0bhSYzeaLG40djsKHAoA

**Related Issues:** 

It looks like NonNullable has had similar issues with other complex types: 
https://github.com/Microsoft/TypeScript/issues/23849
https://github.com/Microsoft/TypeScript/issues/27456

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #30706

This allows a conditional to be related to another conditional via constraint. This means, notably, that conditionals which resolve to `? X : never` should now be assignable to `X`, where `X` is another conditional type.

### CODE REVIEW

**Comment by weswigham** [comment]
@typescript-bot user test this
@typescript-bot perf test this
@typescript-bot run dt
@typescript-bot test this

**Comment by weswigham** [comment]
Every test suite is either green or identical to master - this change seems good.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -15974,19 +15974,23 @@ namespace ts {
...
                         }
-                        const defaultConstraint = getDefaultConstraintOfConditionalType(<ConditionalType>source);
-                        if (defaultConstraint) {
-                            if (result = isRelatedTo(defaultConstraint, target, reportErrors)) {
-                                resetErrorInfo(saveErrorInfo);
-                                return result;
-                            }
+                    }
+                    // while conditionals _can_ be related to one another via normal constraint, as, eg, `A extends B ? O : never` should be assignable to `O`
+                    // when `O` is a conditional (`never` is trivially aissgnable to `O`, as is `O`!).
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: The comment doesn't read right. Delete the leading "while"?

Comment (weswigham) [reply src/compiler/checker.ts]: Sure.

**Pull request merged**